ext/ddtrace_profiling_native_extension/stack_recorder.c

```diff
@@ -148,3 +325,95 @@ static void *call_serialize_without_gvl(void *call_args) {
+
+    // If we got here, slot one was not active, let's try slot two
+
+    error = pthread_mutex_trylock(&state->slot_two_mutex);
+    if (error && error != EBUSY) rb_syserr_fail(error, "Unexpected failure during sampler_lock_active_profile for slot_two_mutex");
+
+    // Slot two is active
+    if (!error) return (struct active_slot_pair) {.mutex = &state->slot_two_mutex, .profile = state->slot_two_profile};
+  }
+
+  // We already tried both multiple times, and we did not succeed. This is not expected to happen. Let's stop sampling.
+  rb_raise(rb_eRuntimeError, "Failed to grab either mutex in sampler_lock_active_profile");
+}
+
+static void sampler_unlock_active_profile(struct active_slot_pair active_slot) {
+  int error = pthread_mutex_unlock(active_slot.mutex);
+  if (error != 0) rb_syserr_fail(error, "Unexpected failure in sampler_unlock_active_profile");
+}
+
+static ddprof_ffi_Profile *serializer_flip_active_and_inactive_slots(struct stack_recorder_state *state) {
+  int error;
+  int previously_active_slot = state->active_slot;
+
+  if (previously_active_slot != 1 && previously_active_slot != 2) {
+    rb_raise(rb_eRuntimeError, "Unexpected active_slot state %d in serializer_flip_active_and_inactive_slots", previously_active_slot);
+  }
+
+  pthread_mutex_t *previously_active = (previously_active_slot == 1) ? &state->slot_one_mutex : &state->slot_two_mutex;
+  pthread_mutex_t *previously_inactive = (previously_active_slot == 1) ? &state->slot_two_mutex : &state->slot_one_mutex;
+
+  // Release the lock, thus making this slot active
+  error = pthread_mutex_unlock(previously_inactive);
+  if (error) rb_syserr_fail(error, "Unexpected failure during serializer_flip_active_and_inactive_slots for previously_inactive");
+
+  // Grab the lock, thus making this slot inactive
+  error = pthread_mutex_lock(previously_active);
+  if (error) rb_syserr_fail(error, "Unexpected failure during serializer_flip_active_and_inactive_slots for previously_active");
+
+  // Update active_slot
+  state->active_slot = (previously_active_slot == 1) ? 2 : 1;
+
+  // Return profile for previously active slot (now inactive)
+  return (previously_active_slot == 1) ? state->slot_one_profile : state->slot_two_profile;
+}
+
+// This method exists only to enable testing Datadog::Profiling::StackRecorder behavior using RSpec.
+// It SHOULD NOT be used for other purposes.
+static VALUE _native_active_slot(DDTRACE_UNUSED VALUE _self, VALUE recorder_instance) {
+  struct stack_recorder_state *state;
+  TypedData_Get_Struct(recorder_instance, struct stack_recorder_state, &stack_recorder_typed_data, state);
+
+  return INT2NUM(state->active_slot);
+}
+
+// This method exists only to enable testing Datadog::Profiling::StackRecorder behavior using RSpec.
+// It SHOULD NOT be used for other purposes.
+static VALUE _native_is_slot_one_mutex_locked(DDTRACE_UNUSED VALUE _self, VALUE recorder_instance) { return test_slot_mutex_state(recorder_instance, 1); }
```

Comment:
How about adding `test` to the name of these methods, since they are to be used only for testing.

Reply:
Or you could create a decorator macro 
(e.g. `DDTRACE_TEST`), and do:
```c
DDTRACE_TEST static VALUE _native_is_slot_one_mutex_locked(
```

Reply:
> DDTRACE_TEST static VALUE _native_is_slot_one_mutex_locked(
> You can even remove the leading _ if it's clearly labeled like this^

I'll address the test feedback next, but I wanted to put in a word on the `_native_` prefix: The `_native_` prefix is something I've been trying to keep in every method that is accessible from the Ruby side. This is so that when you see something going from Ruby to C you know clearly "ooh I'm going to C" rather than a "wait, where is this method defined, I'm not seeing it".

E.g.:

```ruby
      def validate_exporter(exporter_configuration)
        self.class._native_validate_exporter(exporter_configuration)
      end
```

versus calling the native method `validate_exporter` and making it a class method.

> How about adding test to the name of these methods, since they are to be used only for testing.

I think you had commented on a previous PR that you're concerned about making it clear these are test methods as well.

Since there may be multiple concerns here, I'm curious:
* Do you think having a decorator or a prefix (perhaps `_test_only_`) would be enough to make it clear that these methods are not to be used?
* What do you see a potential `DDTRACE_TEST` meaning?

Do you think it's worth looking into not compiling or not even shipping this code? Two potential options I could explore would be to `#ifdef` these out by default in a production release, or even just move testing code to `test_foo.c` files and then exclude `test_*.c` files from the release process.

On one hand having the test code may come in handy if we need to debug something and ask the customers to run some custom piece of code to log or expose something. On the other hand, maybe YAGNI so shipping and compiling less code is worth it.

I don't have strong opinions formed on this yet, so very open to your ideas.

Reply:
Ruby's "new" thing for calling native stuff read pretty neatly: https://github.com/ruby/ruby/blob/3cad54a246e6f292cc58d10ff002525edc6f404a/trace_point.rb#L97-L99

I don't think the code in this PR that much different, tbh.

I'm happy with some sort of indication that a method call is native.

I could suggest this:

```ruby
      def validate_exporter(exporter_configuration)
        Native.validate_exporter(exporter_configuration)
      end
```
But it's probably no different enough to warrant such namespace change.

Reply:
> On one hand having the test code may come in handy if we need to debug something and ask the customers to run some custom piece of code to log or expose something. On the other hand, maybe YAGNI so shipping and compiling less code is worth it.

If this is ever a possibility, than I'd say always ship it and compile it for the user.

> Do you think having a decorator or a prefix (perhaps _test_only_) would be enough to make it clear that these methods are not to be used?
> 
> What do you see a potential DDTRACE_TEST meaning?

I see these two as the same thing, at least in C-land: they label a method as test.
But because these methods are exported to Ruby, I think having something that's visible in Ruby-land makes sense.

How about namespacing the test methods:
* `StackRecorder::Test._native_active_slot`
* `StackRecorder::Test.active_slot`
* `StackRecorder::NativeTest.active_slot`
* `StackRecorder::Native._test_active_slot`

Looking at these, I like the separation of test methods from the main "production" methods/features that live in the `StackRecorder` class.

I don't have a super strong opinion on which option to choose here, the main thing I ask for is a clear indication that some methods are `test` methods that's visible in Ruby-land.

I want to make a note about the `StackRecorder::Test.active_slot` suggestion specifically: this suggestion does not have any native indication, but methods live in a separate `Test` namespace. To me, I'm not as confused regarding where `active_slot` comes from if the whole `Test` module is documented as being comprised of native methods. Not a strong opinion here, just wanted to point out that Module/Class separation for such test and native methods is another approach.

Reply:
Thanks for the feedback. I like the idea of namespacing stuff, although I think I'm still not ready to give up my loved `_native_` prefix.

This makes me lean towards `StackRecorder::Test._native_active_slot` (or `Testing`?).

Since I'm using this pattern elsewhere, if you agree, I'll submit a separate PR (next week probably) applying this refactoring to every test method on the profiling native extension.

Reply:
(Out of curiosity, I was peeking at how Ruby implemented the `Primitive` thing, and seeing if it would be usable for extensions but I wasn't completely sure, I think it's not available. Of course we could just have a `Primitive` class, but that's not what the VM does for its on primitive stuff -- there's actually no `Primitive` in there).

Reply:
A follow-up PR sounds good. 👍

Reply:
I think the Primitive class/middle from the Ruby language repo is build time magic.